### PR TITLE
[KubeRay][NodeProvider] Ignore pods with deletionTimestamp.

### LINKED ISSUE #24514: [kuberay][autoscaler] Filter pods marked for termination out of pod lists.
### What happened + What you expected to happen

When processing info about currently active ("non_terminated") Ray pods the KubeRay node provider should not take into account pods that have been marked for termination. The KubeRay node provider's non_terminated_nodes, create_node, and terminate_nodes methods need to be modified to filter out pods that have been marked for termination, whenever we GET the pods, namely [here](https://github.com/ray-project/ray/blob/70d3bfcf9c9f0a702fb5db6bbc73d5c2ec861586/python/ray/autoscaler/_private/kuberay/node_provider.py#L245) [here](https://github.com/ray-project/ray/blob/70d3bfcf9c9f0a702fb5db6bbc73d5c2ec861586/python/ray/autoscaler/_private/kuberay/node_provider.py#L245) and [here](https://github.com/ray-project/ray/blob/70d3bfcf9c9f0a702fb5db6bbc73d5c2ec861586/python/ray/autoscaler/_private/kuberay/node_provider.py#L268).

A Ray pod is marked for termination when the KubeRay operator issues the request to delete the pod. After that, it may take several seconds before the pod's processes are stopped and the pod is deleted; but that's basically K8s's business -- the autoscaler doesn't need to know about that. The autoscaler does need to know that the operator has initiated the process of removing the pod. If we don't filter out pods marked for termination, we may see thrashing events where extra pods are terminated and then re-created.

### Versions / Dependencies

KubeRay node provider in current Ray master.

### Reproduction script

No specific repro besides anecdotal mentions of unexpected thrashing.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Related issue number

Closes https://github.com/ray-project/ray/issues/24514 by filtering out pods with `metadata.deletionTimestamp` set in the `KuberayNodeProvider`.

Adds some e2e test logic to confirm reasonable behavior when Ray worker pod termination hangs.

Also, a bit of code cleanup -- defining constants, adding to doc strings, etc.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
  E2E tests.

### CODE REVIEW

File: python/ray/autoscaler/_private/kuberay/node_provider.py [hunk 1]
```diff
@@ -191,6 +191,33 @@ def _get(self, path: str) -> Dict[str, Any]:
...
+        """
+        label_filters = to_label_selector(
+            {
+                "ray.io/cluster": self.cluster_name,
+            }
+        )
+        data = self._get("pods?labelSelector=" + requests.utils.quote(label_filters))
+        result = []
+        for pod in data["items"]:
+            if "deletionTimestamp" in pod["metadata"]:
```
Comment (sriram-anyscale) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:212]: Add a comment here about where 'deletionTimestamp' comes from (essentially that it is what k8s uses)

**Comment by sriram-anyscale** [comment]
Looked through the code and the tests. All look good. Included one optional suggestion

File: python/ray/autoscaler/_private/kuberay/node_provider.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:218]: Added comment + link to K8s docs.

**Review by sriram-anyscale — APPROVED** [review]

**Comment by DmitriGekhtman** [comment]
RLlib failure and hanging Mac tests are unrelated.

**Pull request merged**